Improve deeplink tracking

planarvoid opened the pull request:
Fixes #14647 

I'm sorry for the size of this PR. It unexpectedly grew over the time with the changing requirements. What it contains is:
- strip the URL tracked with each deep link click of any data containing any identifiers
- add `source` to each event - tracking where the link came from - email, banner or link
- add `source_info` to each event - additional information - in our case only `login_reason` from the `create_site` links

In order to achieve this I've extracted a `DeepLinkHandler` interface and create a provider for all the handlers. I think the resulting structure is neat and simplifies the `ViewModel` a lot. Let me know what you think! I've tried to cover everything with unit tests.

For the `read` URLs from the `ReaderLinkHandler` that are now redirected to the `ReaderPostPagerActivity` - these are now tracked twice. I think that's the cleanest approach at the moment. Both of these have a different `action` so it should be easy to differentiate between these two events. Let me know what you think about this approach!

To test:
- Try various deeplinks and app links (`wordpress.com`, `public-api.wordpress.com` tracking links, `wordpress://...` app links) 
- Check the `deep_linked` event in Logcat
- Notice it has the following properties:
  - intent_host - it was there before and is the host of the original link
  - source
  - url (which now contains the stripped URL)
  - source_info (only there for the `Create a site` links if `login_reason` is present in the original tracking URL)

## Regression Notes
1. Potential unintended areas of impact
- All the deeplinks handling

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- I've tested all of the deeplinks I could think of and covered everything with unit tests

3. What automated tests I added (or what prevented me from doing so)
- I've added a ton of unit tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

planarvoid committed:
Add new deeplink tracking fields and fix tests

planarvoid committed:
Introduce DeepLinkHandler.kt interface

planarvoid committed:
Implement new interface in NotificationsLinkHandler.kt

planarvoid committed:
Use DeepLinkHandler interface in EditorLinkHandler.kt

planarvoid committed:
Add additional tracking to ReaderPostPagerActivity

planarvoid committed:
Implement DeepLinkHandler.kt in ReaderLinkHandler.kt

planarvoid committed:
Use DeepLinkHandler interface in StartLinkHandler.kt

planarvoid committed:
Use DeepLinkHandler interface in StatsLinkHandler.kt

planarvoid committed:
Merge branch 'develop' into feature/14647-improve-deeplink-tracking

planarvoid committed:
Use DeepLinkHandler interface in PagesLinkHandler.kt

planarvoid committed:
Move DeepLink handlers into a separate class

planarvoid committed:
Move tracking of reader links to the DeepLinkTrackingUtils.kt

planarvoid committed:
Cleanup ReaderLinkHandler.kt

planarvoid committed:
Fix checkstyle issues

planarvoid committed:
Fix ktlint issue

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/deeplinks/EditorLinkHandler.kt:
```diff
@@ -50,11 +54,44 @@ class EditorLinkHandler
+    override fun stripUrl(uri: UriWrapper): String {
+        return if (uri.host == POST_PATH) {
+            // Transforms wordpress://post?blogId=798&postId=1231 to wordpress://post?blogId=blogId&postId=postId
+            buildString {
```
TIL about `buildString`!

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/deeplinks/NotificationsLinkHandler.kt:
```diff
@@ -6,24 +6,28 @@ import org.wordpress.android.util.UriWrapper
     }
 
+    override fun stripUrl(uri: UriWrapper): String {
+        return uri.host + "/" + uri.pathSegments.firstOrNull()
```
This is missing the handling of the `wordpress://` scheme. Also, when the path segment is null, like on `wordpress://notifications`, we get an actual `null` appended to the end of the URL. Maybe we should handle this like we do in `StatsLinkHandler`?

planarvoid replied:
ah you're right 👍 . I missed this usecase, it's fixed in 
0fe5156

renanferrari reviewed (commented):
Thanks a lot for the changes, the code has definitely improved!

I've tested with a bunch of different links and everything seems to be working as expected, except for one case, which I've detailed below.

Other than that, I thought maybe we should move all link handlers and its related classes (`DeepLinkHandler` and `DeepLinkHandlers`) to their own package inside `deeplinks`, what do you think?

planarvoid committed:
Handle applinks in NotificationsLinkHandler.kt

planarvoid committed:
Move handlers to a new package

planarvoid commented:
thanks for the review @renanferrari ! I've fixed the issue and also moved the handlers into an inner package in dee1f57, I think it makes a lot of sense 👍

renanferrari reviewed (approved):
Thanks for the changes!

LGTM :shipit:
